### ISSUE TITLE
pandas dataframes now automatically converted to `Dict`?

### ISSUE THREAD

**Issue by ExpandingMan** [JuliaPy/PyCall.jl#355]
I couldn't find any information about this, so I thought I'd file this issue.

As of 1.8.0, pandas dataframes are automatically converted to Julia `Dict`s.  I'm by no means sure this is a bug, perhaps it's desired behavior?  The dict has keys equal to the columns, and values equal to `Dict`s with keys equal to row numbers and the values equal to the actual dataframe element values.

This behavior isn't exactly *bad*, but there are a couple of problems I see.  First, the `Dict`s are all `Dict{Any,Any}`, so inferring types is a big pain.  Granted, the situation wasn't great when pulling from a pandas dataframe, but at least in that case any column that is a numpy array gets converted to it's proper Julia type just fine.  Second, shouldn't the columns be `Vector`s instead of `Dict`s?

I've had some methods to convert pandas dataframes for a while, the major problem with them is that properly converting types can be a big headache, due to the fact that there isn't any proper metadata on columns of Python objects, and also pandas doesn't have any uniform, cogent way of dealing with missing values.

**Comment by stevengj** [comment]
Basically, PyCall uses `Dict` for types that report themselves as "mapping" objects.

The general pitfall with automatically converting Python types to Julia is that sometimes PyCall doesn't pick the type you want.  The tradeoff is that, in the common case where it does pick the type you want, it is a lot more convenient.

You can always manually specify what type you want, or inhibit conversion entirely, though, e.g. by using `o["member"]` instead of `o[:member]` or `pycall(o["member"], PyObject, args...)` instead of `o[:member](args...)`.   You can also register custom converters for specific Python types.

**Comment by ExpandingMan** [comment]
Ah, got it.  I didn't really feel like this was a proper "issue", but it was new behavior and I didn't realize there was a workaround.

Thanks!

**Comment by stevengj** [comment]
It's not really a "new" behavior, it's just that the PyMapping conversion was broken previously except for `dict`.

**Comment by ExpandingMan** [comment]
I see.  Well that was a pretty major bug because it completely changes the behavior of a huge number of objects!

Thanks for fixing it!

**Comment by malmaud** [comment]
You might also be interested in [Pandas.jl](https://github.com/JuliaPy/Pandas.jl), which adds some quality-of-life conveniences on top of using PyCall with Pandas directly.

**Comment by ExpandingMan** [comment]
Thanks @malmaud, I was aware of that package, it's nice to have.  Unfortunately the Julia DataFrames package isn't in such a good state right now.  Some of what I'm doing right now involves a lot of missing values, so I've switched over to the DataFrames.jl master, as it has much better performance in that situation because of type stability (basically everything is a `Nullable`).  Discouragingly, the community hasn't quite made up its mind yet on how to move forward, and there's limited tools available for using the new `DataFrame`s (I wrote a bunch of quick ones myself).  

The main reason I'm using Python at all right now is to query databases.  The Julia tools for doing that are poorly maintained.  Fortunately, the changes to PyCall have made it much easier to query using `sqlalchemy` and plug the results directly into a Julia `DataFrame`, which turns out to be way faster than building a pandas dataframe in the first place (though so far it's a bit buggy)!

Anyway, thanks again.
